### PR TITLE
Generalize Interface Sorption

### PR DESCRIPTION
**Pull request by simopier**
Update `InterfaceSorptionSievert` into `InterfaceSorption` to make it general:
- The exponent in the sorption law can now be user-defined
- Update documentation
- Edit existing tests and add tests for Henry's law
- Update name of test folder to reflect generalization

(close. #156)

### CODE REVIEW

File: doc/content/source/interfacekernels/InterfaceSorption.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+!syntax description /InterfaceKernels/InterfaceSorption
+
+## Description
+
+The surface concentrations $C_s$ in the solid are related to the partial pressure $P$ (and corresponding gas concentration $C_g$ assuming ideal gas behavior) of the neighboring gas by the interface sorption law:
+\begin{equation}
+C_s = K P^n = K \left(C_g R T \right)^n,
+\end{equation}
+where $R$ is the ideal gas constant in J/mol/K, $T$ is the temperature in K, and $K$ is the solubility, and $n$ is the exponent of the sorption law. $K$ which is described as an Arrhenius equation:
```
Comment (cticenhour) [inline doc/content/source/interfacekernels/InterfaceSorption.md:11]:
```suggestion
where $R$ is the ideal gas constant in J/mol/K, $T$ is the temperature in K, $K$ is the solubility, and $n$ is the exponent of the sorption law. $K$ is described as an Arrhenius equation:
```

File: doc/content/source/interfacekernels/InterfaceSorption.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+\begin{equation}
+K = K_0 \exp \left(\frac{-E_a}{RT}\right),
+\end{equation}
+where $K_0$ is the pre-exponential constant, and $E_a$ is the activation energy in J/mol. The units of $K$ and $K_0$ depend on the units of $C_s$ and $P$, as well as the value of $n$.
+
+The value of $n$ determines the sorption law. For example:
+- $n = 1$ corresponds to Henry's law
+- $n=1/2$ corresponds to Sievert's law.
```
Comment (cticenhour) [inline doc/content/source/interfacekernels/InterfaceSorption.md:19]:
```suggestion
- $n = 1/2$ corresponds to Sievert's law.
```

File: doc/content/source/interfacekernels/InterfaceSorption.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+To balance the mass flux across the gap, a second interfacial condition is given as:
+
+\begin{equation}
+D_s \nabla C_s \cdot \mathbf{n_s} + D_g \nabla C_g \cdot \mathbf{n_g} = 0,
+\label{eqn:flux}
+\end{equation}
+where $D_s$ and $D_g$ are the diffusivities in the solid and the gas, respectively, and $\mathbf{n_s}$ and $\mathbf{n_g}$ are the normals at the interface. Two methods are available to enforce flux conditions at the interface. By default, `InterfaceSorption` applies the flux at the primary side of the interface to the residual on the neighbor side of the interface. Using a kernel such as [MatDiffusion](/MatDiffusion.md) on the neighbor block adds the flux at the neighbor side of the interface to the residual, forming the full flux balance condition shown in the equation above.
+
+!alert warning title=Enforcing flux conditions using the default, non-penalty method
+When using the default, non-penalty method to enforcing flux conditions at the interface, a kernel such as [MatDiffusion](/MatDiffusion.md) must be applied to the neighbor block to form the correct residual.
```
Comment (cticenhour) [inline doc/content/source/interfacekernels/InterfaceSorption.md:32]:
```suggestion
When using the default, non-penalty method to enforce flux conditions at the interface, a kernel such as [MatDiffusion](/MatDiffusion.md) must be applied to the neighbor block to form the correct residual.
```

File: test/tests/interfacekernels/InterfaceSorption/tests [hunk 4]
```diff
@@ -0,0 +1,130 @@
+#These tests currently only run in replicated mesh mode, since the BreakMeshByBlockGenerator
+#is currently only supported for replicated meshes
+[Tests]
+  design = 'InterfaceSorption.md'
+  issues = '#50 #0000000'
```
Comment (cticenhour) [inline test/tests/interfacekernels/InterfaceSorption/tests:5]:
```suggestion
  issues = '#50'
```

File: test/tests/interfacekernels/InterfaceSorption/tests [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Sievert law in isothermal conditions.'
+    type = 'Exodiff'
+    input = 'interface_sorption.i'
+    cli_args = 'Outputs/file_base=interface_sorption_Sievert_out'
+    exodiff = 'interface_sorption_Sievert_out.e'
+    abs_zero = 1e-7
+    mesh_mode = REPLICATED
+  []
+  [Sievert_non_ad_penalty]
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Sievert law in isothermal conditions using penalty-enforced flux balance.'
```
Comment (cticenhour) [inline test/tests/interfacekernels/InterfaceSorption/tests:16]:
```suggestion
    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Sievert law in isothermal conditions using a penalty-enforced flux balance.'
```

File: test/tests/interfacekernels/InterfaceSorption/tests [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                Postprocessors/flux_inner/type=ADSideDiffusiveFluxIntegral
+                Postprocessors/flux_outer/type=ADSideDiffusiveFluxIntegral
+                Outputs/file_base=interface_sorption_Sievert_out'
+    exodiff = 'interface_sorption_Sievert_out.e'
+    prereq = Sievert_non_ad
+    abs_zero = 1e-7
+    mesh_mode = REPLICATED
+  []
+  [Sievert_ad_penalty]
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Sievert law in isothermal conditions using automatic differentiation and penalty-enforced flux balance.'
```
Comment (cticenhour) [inline test/tests/interfacekernels/InterfaceSorption/tests:47]:
```suggestion
    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Sievert law in isothermal conditions using automatic differentiation and a penalty-enforced flux balance.'
```

File: test/tests/interfacekernels/InterfaceSorption/tests [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Henry law in isothermal conditions.'
+    type = 'Exodiff'
+    input = 'interface_sorption.i'
+    cli_args = 'InterfaceKernels/interface/n_sorption=1 Outputs/file_base=interface_sorption_Henry_out'
+    exodiff = 'interface_sorption_Henry_out.e'
+    abs_zero = 1e-7
+    mesh_mode = REPLICATED
+  []
+  [Henry_non_ad_penalty]
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Henry law in isothermal conditions using penalty-enforced flux balance.'
```
Comment (cticenhour) [inline test/tests/interfacekernels/InterfaceSorption/tests:77]:
```suggestion
    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Henry law in isothermal conditions using a penalty-enforced flux balance.'
```

File: test/tests/interfacekernels/InterfaceSorption/tests [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                Postprocessors/flux_inner/type=ADSideDiffusiveFluxIntegral
+                Postprocessors/flux_outer/type=ADSideDiffusiveFluxIntegral
+                Outputs/file_base=interface_sorption_Henry_out'
+    exodiff = 'interface_sorption_Henry_out.e'
+    prereq = Henry_non_ad
+    abs_zero = 1e-7
+    mesh_mode = REPLICATED
+  []
+  [Henry_ad_penalty]
+    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Henry law in isothermal conditions using automatic differentiation and penalty-enforced flux balance.'
```
Comment (cticenhour) [inline test/tests/interfacekernels/InterfaceSorption/tests:110]:
```suggestion
    requirement = 'The system shall have the capability to enforce interfacial conditions based on the Henry law in isothermal conditions using automatic differentiation and a penalty-enforced flux balance.'
```

**Review by cticenhour — CHANGES_REQUESTED** [review]

File: include/interfacekernels/InterfaceSorption.h [hunk 9]
```diff
@@ -36,6 +36,9 @@ class InterfaceSorptionSievertTempl : public InterfaceKernelParent<is_ad>
   const Real _Ea;
   ///@}
 
+  /// Exponent of the Sorption law
```
Comment (simopier) [inline include/interfacekernels/InterfaceSorption.h:39]:
```suggestion
  /// Exponent of the sorption law
```

[the remark this answers is not in the document]
Comment (simopier) [reply include/interfacekernels/InterfaceSorption.h]: Made it not capitalized since it's a noun. Good catch!

**Review by cticenhour — APPROVED** [review]

**Pull request merged**